arch/cpu/native/Makefile.native

```diff
@@ -20,6 +20,25 @@ endif
 CFLAGSNO = -Wall -g -I/usr/local/include $(CFLAGSWERROR)
 CFLAGS  += $(CFLAGSNO)
 
+### Are we building with code size optimisations?
+SMALL ?= 0
+
+# The optimizations on native platform cannot be enabled in GCC versions less than 7.2
+GCC_VERSION := $(shell gcc --version | grep ^gcc | sed 's/^.* //g' | cut -b1-3)
+ifeq ($(shell expr $(GCC_VERSION) \>= 7.2), 1)
+	NATIVE_CAN_OPTIIMIZE = 1
+else
+	NATIVE_CAN_OPTIIMIZE = 0
+endif
+
+ifeq ($(NATIVE_CAN_OPTIIMIZE),1)
+  ifeq ($(SMALL),1)
+    CFLAGS += -Os
+  else
+    CFLAGS += -O2
+  endif
+endif
+
```

Comment:
Suggest putting the gcc version retrieval inside the linux conditional below (it fails on macos). On macos setting NATIVE_CAN_OPTIIMIZE to 1 unconditionally should be OK

Reply:
Are you sure about the latter?

I think it would be good to design a GCC_VERSION makefile variable that works on both. Does `gcc -dumpfullversion -dumpversion` work on mac? Is it actually a GCC version format issue or shell script issue?

Reply:
```
$ gcc -dumpfullversion -dumpversion
4.2.1
```
```
$ gcc --version
Configured with: --prefix=/Library/Developer/CommandLineTools/usr --with-gxx-include-dir=/usr/include/c++/4.2.1
Apple LLVM version 10.0.0 (clang-1000.10.44.2)
Target: x86_64-apple-darwin17.7.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
```

Draw your own conclusions... :)

But I don't think we need to try to support mega old macos / XCode command line tool versions

Reply:
How about this?

Reply:
Yeah that looks good. Tested over here.

Reply:
By the way the "Configured with:" line seems to be going to stderr, you may want to consider

```
GCC_IS_CLANG := $(shell gcc --version 2> /dev/null | grep clang)
```